test/integration/client-side-operations-timeout/node_csot.test.ts

```diff
@@ -804,6 +805,182 @@ describe('CSOT driver tests', metadata, () => {
+
+          await internalClient.db().admin().command(failpoint);
+          cs = client.db('db').collection('coll').watch([], { timeoutMS: 120 });
+          errorIter = on(cs, 'error');
+          cs.on('change', () => {
+            // Add empty listener just to get the change stream running
+          });
+        });
+
+        it('emits an error event', metadata, async function () {
+          const [err] = (await errorIter.next()).value;
+          expect(data).to.have.lengthOf(0);
+          expect(err).to.be.instanceof(MongoOperationTimeoutError);
+        });
+
+        it('continues emitting change events', metadata, async function () {
+          const err = (await errorIter.next()).value[0];
+          expect(err).to.be.instanceof(MongoOperationTimeoutError);
+
+          await once(cs, 'resumeTokenChanged');
+
+          await client.db('db').collection('coll').insertOne({ x: 1 });
+
+          const [change] = await once(cs, 'change');
+          expect(change).to.have.ownProperty('operationType', 'insert');
+        });
+
+        it('does not close the change stream', metadata, async function () {
+          const [err] = (await errorIter.next()).value;
+          expect(err).to.be.instanceof(MongoOperationTimeoutError);
+
+          expect(cs.closed).to.be.false;
+        });
+
+        it('attempts to create a new change stream cursor', metadata, async function () {
+          await errorIter.next();
+          let aggregates = commandsStarted
+            .filter(x => x.commandName === 'aggregate')
+            .map(x => x.command);
+          expect(aggregates).to.have.lengthOf(1);
+
+          await once(cs, 'resumeTokenChanged');
+
+          aggregates = commandsStarted
+            .filter(x => x.commandName === 'aggregate')
+            .map(x => x.command);
+
+          expect(aggregates).to.have.lengthOf(2);
+
+          expect(aggregates[0].pipeline).to.deep.equal([{ $changeStream: {} }]);
+          expect(aggregates[1].pipeline).to.deep.equal([
+            { $changeStream: { resumeAfter: cs.resumeToken } }
+          ]);
+        });
+      });
+
+      context('when the resume attempt times out', function () {
+        const failpoint: FailPoint = {
+          configureFailPoint: 'failCommand',
+          mode: { times: 2 }, // timeout the getMore, and the aggregate
+          data: {
+            failCommands: ['getMore', 'aggregate'],
+            blockConnection: true,
+            blockTimeMS: 130
+          }
+        };
+
+        beforeEach(async function () {
+          cs = client.db('db').collection('coll').watch([], { timeoutMS: 120 });
+          const _changePromise = once(cs, 'change');
+          await once(cs.cursor, 'init');
+
+          await internalClient.db().admin().command(failpoint);
+        });
+
+        it('emits an error event', async function () {
+          let [err] = await once(cs, 'error'); // getMore failure
+          [err] = await once(cs, 'error'); // aggregate failure
```

Comment:
Do we want to check something about each error instead of just the last?

I am surprised we emit more than one error event. I supposed that is fine but is there a close event that follows all of these?

Reply:
Good catch here. It does seem that there are two close events being emitted here; one for the initial timeout error that will be resumed and the second for the aggregate that would have been the resume attempt.

Would we prefer not to emit an error event on the getMore's timeout? We could instead add a custom timeout event that reports the timeout error but doesn't cause the change stream to emit the close event.

Reply:
Post the link here when you have it and we can resolve TY

Reply:
Filed: NODE-6434